### ISSUE TITLE
Renewed junctionTable id's on every parent update

### ISSUE THREAD

**Issue by MarijnMensinga** [balderdashy/waterline#1144]
Hi,

**The issue**
This last day I have been debugging waterline to find out why records in a junction table get new id's on every update of any of it's two parent tables.

**The reason (tl;dr)** 
The reason of new id behaviour is that after updating one of the two parent records ([Code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L207)) all nested associations get 'updated' ([Code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L222)); which means that all collection associations get removed and then re-added.

**The reason (in depth)**
While debugging I found out that a change set of associations is maintained by waterline while using objects. So when ``model.save()`` is called this is what happens:

1. update parent model [code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L83)
  * update the model [code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L207)
  * remove all associations [code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/utils/nestedOperations/update.js#L185)
  * re-add all associations [code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/utils/nestedOperations/update.js#L191)
1. if there are new associations: add them [code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L120)
2. if there are associations to be removed: remove them [code] (https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L138)

From this flow I deducted I should not be needed to remove and re-add the associations, so I created some usecases in which I skipped that part of the code (``UpdateNestedAssociations(..)``,[code](https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L222)). The results where as expected, adding and removing kept working but no new id's were introduced. So I'm starting to think that the ``UpdateNestedAssociations(..)`` is not needed as higher-level functions successfully deal with the adding and removing of associations.

**My questions**
1. Is it expected / desired behaviour that records in a junction table get new id's on every parent update? (I'm assuming it is not, because of key space exhaustion)
2. Is the ``UpdateNestedAssociations()`` necessary anyway? If so, can someone explain why?

Thanks in advance!
Marijn

**Comment by devinivy** [comment]
@MarijnMensinga you're right about this flow.  However, it should only perform nested updates if `Collection.update()` is supplied values that imply setting of associations.  `Collection.update()` will _reset_ associations to the parent record when associated records are specified to match the exact list of associated records provided in the `Collection.update()` values object.  Though `Model.save()` defers to `Collection.update()`, I believe that it does so in a way that does not trigger the nested  updates.

We have identified that this flow can be better optimized to minimize roundtrips to the db and data-touching.

Do you mind sharing some specific code where you're seeing this behavior?

**Comment by MarijnMensinga** [comment]
Yes I have an example.

**Agent**
```javascript
module.exports = {
  tableName: 'agent',
  migrate: 'safe',
  autoCreatedAt: false,
  autoUpdatedAt: false,
  schema	: true,
  attributes: {
    username: {
      type: 'string',
      required: true,
      maxLength: 50
    },
	groups: {
      collection: 'group',
      via: 'agents',
      through: 'group_agent'
    }
  }	
};	
```

**Group**
```javascript
module.exports = {
  tableName: 'group',
  migrate: 'safe',
  autoCreatedAt: false,
  autoUpdatedAt: false,
  schema	: true,
  attributes: {
    name: {
      type: 'string',
      required: true,
      maxLength: 50
    },
    agents: {
      collection: 'agent',
      via: 'groups',
      dominant:true,
      through: 'group_agent'
    }
};	
```

**Group_Agent** _Junction table_
```javascript
module.exports = {
  tableName: 'group__agent',
  migrate: 'safe',
  junctionTable: true,
  autoCreatedAt: false,
  autoUpdatedAt: false,
  schema	: true,
  attributes: {
    agents: {
      columnName: 'agents',
      type: 'integer',
      foreignKey: true,
      references: 'agent',
      on: 'id',
      via: 'groups',
      groupBy: 'agent'
    },
    groups: {
      columnName: 'groups',
      type: 'integer',
      foreignKey: true,
      references: 'group',
      on: 'id',
      via: 'agents',
      groupBy: 'group'
    }
  }
};
```

**Test**
Test 1 and Test 2 both result in new PK's for all records belonging to group 1 within the junction table
```javascript
var group,agent;
async.waterfall([
	function fetchGroup(done){
		Group.findOne(1).populate('agents').exec(function(err,dbGroup){
			group = dbGroup;
			done(err);
		});
	},
	function fetchAgent(done){
		Agent.findOne(4,function(err,dbAgent){
			agent = dbAgent;
			done(err);
		});
	},
	function test1(done){
		group.name = 'ICT group 2';
		group.save(function(err,dbGroup){ //This triggers new id's
			group = dbGroup
			done(err);
		});
	},
	function test2(done){
		group.agents.add(agent);
		group.save(function(err,dbGroup){ //This triggers new id's
			group = dbGroup
			done(err);
		});
	}
],function(err){
  //done
});
```

**Comment by particlebanana** [comment]
This is because of the "syncing" that takes place. We need to address this and turn it off or fix the `save` method to change this behavior. Remove the `populate` and this will be taken care of in the example above.

**Comment by sailsbot** [comment]
Thanks for posting, @MarijnMensinga.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
